### PR TITLE
BaseTarget calculation algorithm fixing the blocktimes

### PR DESCRIPTION
**Pull request by asolovyov**
See details here:

  * http://wiki.nxtcrypto.org/wiki/Whitepaper:Nxt#Base_Target_Value
  * https://nxtforum.org/proof-of-stake-algorithm/basetarget-adjustment-algorithm/
  * https://nxtforum.org/nxt-improvement-proposals/fixing-the-blocktimes

### CODE REVIEW

File: build-submodules.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
```
Comment (catena2w) [inline build-submodules.sh]: It's better to write sbt commands for such a tasks

File: scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala [hunk 2]
```diff
@@ -10,11 +10,10 @@ import scorex.utils.{NTP, ScorexLogging}
 
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
-import scala.concurrent.duration._
 import scala.util.{Failure, Try}
 
 
-class NxtLikeConsensusModule(AvgDelay: Long = 5.seconds.toMillis)
+class NxtLikeConsensusModule(AvgDelayInSeconds: Long = 60)
```
Comment (catena2w) [inline scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala]: Let's update version on line 23 to `val version = 2: Byte`

File: scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala [hunk 3]
```diff
@@ -95,27 +97,53 @@ class NxtLikeConsensusModule(AvgDelay: Long = 5.seconds.toMillis)
     } else Future(None)
   }
 
+  private def effectiveBalance[TT](account: Account)(implicit transactionModule: TransactionModule[TT]) =
+    transactionModule.blockStorage.state.asInstanceOf[BalanceSheet].balanceWithConfirmations(account.address, EffectiveBalanceDepth)
```
Comment (catena2w) [inline scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala]: Line length should be <= 120 letters

File: scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala [hunk 4]
```diff
@@ -95,27 +97,53 @@ class NxtLikeConsensusModule(AvgDelay: Long = 5.seconds.toMillis)
...
-    val prevBt = BigInt(lastBlockData.baseTarget)
-    val t0 = bounded(prevBt * eta / AvgDelay, prevBt / 2, prevBt * 2)
-    bounded(t0, 1, Long.MaxValue).toLong
+    BigInt(1, calcGeneratorSignature(lastBlockData, generator).take(8).reverse)
+
+  /**
+    * BaseTarget calculation algorithm fixing the blocktimes.
+    *
+    */
+  private def calcBaseTarget[TT](prevBlock: Block, timestamp: Long) (implicit transactionModule: TransactionModule[TT]): Long = {
```
Comment (catena2w) [inline scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala]: too long line

File: scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala [hunk 5]
```diff
@@ -95,27 +97,53 @@ class NxtLikeConsensusModule(AvgDelay: Long = 5.seconds.toMillis)
...
+    if (height % 2 == 0) {
+      val lastBlocks: Seq[Block] = transactionModule.blockStorage.history.lastBlocks(3)
+      val block = lastBlocks.head
+      val blocktimeAverage = (timestamp - block.timestampField.value) / lastBlocks.size
+      val bt =
+        if (blocktimeAverage > AvgDelayInSeconds) {
+          (prevBaseTarget * Math.min(blocktimeAverage, MaxBlocktimeLimit)) / AvgDelayInSeconds
+        }
+        else {
+          prevBaseTarget - prevBaseTarget * BaseTargetGamma * (AvgDelayInSeconds - Math.max(blocktimeAverage, MinBlocktimeLimit)) / (AvgDelayInSeconds * 100)
```
Comment (catena2w) [inline scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala]: too long line

File: scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala [hunk 6]
```diff
@@ -95,27 +97,53 @@ class NxtLikeConsensusModule(AvgDelay: Long = 5.seconds.toMillis)
...
+    val prevBaseTarget = consensusBlockData(prevBlock).baseTarget
+    if (height % 2 == 0) {
+      val lastBlocks: Seq[Block] = transactionModule.blockStorage.history.lastBlocks(3)
+      val block = lastBlocks.head
+      val blocktimeAverage = (timestamp - block.timestampField.value) / lastBlocks.size
+      val bt =
+        if (blocktimeAverage > AvgDelayInSeconds) {
+          (prevBaseTarget * Math.min(blocktimeAverage, MaxBlocktimeLimit)) / AvgDelayInSeconds
+        }
+        else {
```
Comment (catena2w) [inline scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala]: Use else on the same line with bracket `} else {`

File: scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala [hunk 7]
```diff
@@ -95,27 +97,53 @@ class NxtLikeConsensusModule(AvgDelay: Long = 5.seconds.toMillis)
...
+
+  /**
+    * BaseTarget calculation algorithm fixing the blocktimes.
+    *
+    */
+  private def calcBaseTarget[TT](prevBlock: Block, timestamp: Long) (implicit transactionModule: TransactionModule[TT]): Long = {
+    val height = transactionModule.blockStorage.history.heightOf(prevBlock).get
+    val prevBaseTarget = consensusBlockData(prevBlock).baseTarget
+    if (height % 2 == 0) {
+      val lastBlocks: Seq[Block] = transactionModule.blockStorage.history.lastBlocks(3)
```
Comment (catena2w) [inline scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala]: You can use `transactionModule.blockStorage.history.averageDelay(prevBlock, 3)` here

File: scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala [hunk 8]
```diff
@@ -95,27 +97,53 @@ class NxtLikeConsensusModule(AvgDelay: Long = 5.seconds.toMillis)
...
+      val block = lastBlocks.head
+      val blocktimeAverage = (timestamp - block.timestampField.value) / lastBlocks.size
+      val bt =
```
Comment (catena2w) [inline scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala]: Type `if` on the same line as val, like `val bt = if (blocktimeAverage > AvgDelayInSeconds) {`

File: scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala (hunk 2 above)
Comment (gagarin55) [inline scorex-consensus/src/main/scala/scorex/consensus/nxt/NxtLikeConsensusModule.scala]: AvgDelay will pass from our Waves Application, we don't need to change its default value (5 sec) in Scorex

**Pull request merged**